libs/metrics/in_out.py

```diff
@@ -168,87 +177,41 @@ def can_execute(cls, config, entity):
-    if angle < 180: # in
-        return 1, 0
-    else: # out
-        return 0, 1
-
-def check_intersect(p1, p2, p3, p4):
-    """
-    Check if the line p1-p2 intersects the line p3-p4
-    Args:
-        p1: (x,y)
-        p2: (x,y)
-        p3: (x,y)
-        p4: (x,y)
-
-    Returns:
-        boolean : True if intersection occurred
-    """
-    tc1 = (p1[0] - p2[0]) * (p3[1] - p1[1]) + (p1[1] - p2[1]) * (p1[0] - p3[0])
-    tc2 = (p1[0] - p2[0]) * (p4[1] - p1[1]) + (p1[1] - p2[1]) * (p1[0] - p4[0])
-    td1 = (p3[0] - p4[0]) * (p1[1] - p3[1]) + (p3[1] - p4[1]) * (p3[0] - p1[0])
-    td2 = (p3[0] - p4[0]) * (p2[1] - p3[1]) + (p3[1] - p4[1]) * (p3[0] - p2[0])
-    return tc1 * tc2 < 0 and td1 * td2 < 0
-
-def calc_vector_angle(line1_p1, line1_p2, line2_p1, line2_p2):
-    """
-    Calculate the and return the angle made by two line segments line1(p1)-(p2), line2(p1)-(p2)
-    Args:
-        line1_p1: (x,y)
-        line1_p2: (x,y)
-        line2_p1: (x,y)
-        line2_p2: (x,y)
-
-    Returns:
-        angle : [0, 360)
-    """
-    u = np.array(line_vectorize(line1_p1, line1_p2))
-    v = np.array(line_vectorize(line2_p1, line2_p2))
-    i = np.inner(u, v)
-    n = LA.norm(u) * LA.norm(v)
-    c = i / n
-    a = np.rad2deg(np.arccos(np.clip(c, -1.0, 1.0)))
-    if u[0] * v[1] - u[1] * v[0] < 0:
-        return a
-    else:
-        return 360 - a
-
-def line_vectorize(point1, point2):
-    """
-    Args:
-        point1: (x,y)
-        point2: (x,y)
-
-    Returns:
-        The vector of intersecting the points with a line line(point1)-(point2)
-    """
-    a = point2[0] - point1[0]
-    b = point2[1] - point1[1]
-    return [a, b]
+
+    @classmethod
+    def get_weekly_report(cls, entities: List[str], number_of_weeks: int = 0,
+                          from_date: date = None, to_date: date = None) -> Dict:
+        # The In/Out metric cannot be fully aggregated using "sum"
+        weekly_report_data = cls.generate_weekly_report_data(entities, number_of_weeks, from_date, to_date)
+        report = { "Weeks": [] }
+        for header in cls.csv_headers:
+            report[header] = []
+        for week, week_data in weekly_report_data.items():
+            report["Weeks"].append(week)
+            report["In"].append(sum(week_data["In"]))
+            report["Out"].append(sum(week_data["Out"]))
+            if is_list_recursively_empty(week_data["Summary"]):
+                boundary_name = []
+                weekly_in = []
+                weekly_out = []
+            else:
+                boundary_names, weekly_in, weekly_out = list(zip(*week_data["Summary"]))
+                boundary_name = next(x for x in boundary_names if not is_list_recursively_empty(x))
+                weekly_in = fill_partially_empty_result(weekly_in, 0)
+                weekly_out = fill_partially_empty_result(weekly_out, 0)
+
+            report["Summary"].append([
+                boundary_name,
+                [sum(x) for x in zip(*weekly_in)],
```

Comment:
It is hard to follow this code without info about types of these variables. Is weekly_in not an array of numbers where you could directly do sum(weekly_in)?